python/hail/expr/expression.py

```diff
@@ -76,109 +76,191 @@ def construct_reference(name, type, indices, prefix=None):
-    elif isinstance(e, dict):
-        if len(e) == 0:
-            raise ExpressionException('Cannot convert empty dictionary to expression.')
-        key_cols = []
-        value_cols = []
+    elif isinstance(x, bool):
+        return tbool
+    elif isinstance(x, int):
+        if hl.tint32.min_value <= x <= hl.tint32.max_value:
+            return tint32
+        elif hl.tint64.min_value <= x <= hl.tint64.max_value:
+            return tint64
+        else:
+            raise ValueError("Hail has no integer data type large enough to store {}".format(x))
+    elif isinstance(x, float):
+        return tfloat64
+    elif isinstance(x, str):
+        return tstr
+    elif isinstance(x, Locus):
+        return tlocus(x.reference_genome)
+    elif isinstance(x, Interval):
+        return tlocus(x.reference_genome)
+    elif isinstance(x, Call):
+        return tcall
+    elif isinstance(x, Struct):
+        return tstruct(**{k: impute_type(x[k]) for k in x})
+    elif isinstance(x, tuple):
+        return ttuple(*(impute_type(element) for element in x))
+    elif isinstance(x, list):
+        ts = {impute_type(element) for element in x}
+        unified_type = unify_types_limited(*ts)
+        if not unified_type:
+            raise ExpressionException("Hail does not support heterogeneous arrays: "
+                                      "found list with elements of types {} ".format(list(ts)))
+        return tarray(unified_type)
+    elif isinstance(x, set):
+        ts = {impute_type(element) for element in x}
+        unified_type = unify_types_limited(*ts)
+        if not unified_type:
+            raise ExpressionException("Hail does not support heterogeneous sets: "
+                                      "found set with elements of types {} ".format(list(ts)))
+        return tset(unified_type)
+    elif isinstance(x, dict):
+        kts = {impute_type(element) for element in x.keys()}
+        vts = {impute_type(element) for element in x.values()}
+        unified_key_type = unify_types_limited(*kts)
+        unified_value_type = unify_types_limited(*vts)
+        if not unified_key_type:
+            raise ExpressionException("Hail does not support heterogeneous dicts: "
+                                      "found dict with keys of types {} ".format(list(kts)))
+        if not unified_value_type:
+            raise ExpressionException("Hail does not support heterogeneous dicts: "
+                                      "found dict with values of types {} ".format(list(vts)))
+        return tdict(unified_key_type, unified_value_type)
+    elif x is None:
+        raise ExpressionException("Hail cannot impute the type of 'None'")
+    else:
+        raise ExpressionException("Hail cannot automatically impute type of {}: {}".format(type(x), x))
+
+
+def to_expr(e, dtype=None):
+    if isinstance(e, Expression):
+        if dtype:
+            assert e.dtype == dtype, 'expected {}, got {}'.format(dtype, e.dtype)
```

Comment:
This should be a user-friendly message, not an assert.  For example, Table.parallelize(v, t) might error out here if v isn't of type t.